From: Andreas Gruenbacher <agruen@suse.de>
Subject: [nfsd] Check for read-only exports before setting acls
References: 139411

We must check for MAY_SATTR before setting acls, which includes
checking for read-only exports: the lower-level setxattr operation
that eventually sets the acl cannot check export-level restrictions.

Bug reported by Martin Walter <mawa@uni-freiburg.de>.

Signed-off-by: Andreas Gruenbacher <agruen@suse.de>

Index: linux-2.6.13-SL100_BRANCH/fs/nfsd/nfs2acl.c
===================================================================
--- linux-2.6.13-SL100_BRANCH.orig/fs/nfsd/nfs2acl.c
+++ linux-2.6.13-SL100_BRANCH/fs/nfsd/nfs2acl.c
@@ -107,7 +107,7 @@ static int nfsacld_proc_setacl(struct sv
 	dprintk("nfsd: SETACL(2acl)   %s\n", SVCFH_fmt(&argp->fh));
 
 	fh = fh_copy(&resp->fh, &argp->fh);
-	nfserr = fh_verify(rqstp, &resp->fh, 0, MAY_NOP);
+	nfserr = fh_verify(rqstp, &resp->fh, 0, MAY_SATTR);
 
 	if (!nfserr) {
 		nfserr = nfserrno( nfsd_set_posix_acl(
Index: linux-2.6.13-SL100_BRANCH/fs/nfsd/nfs3acl.c
===================================================================
--- linux-2.6.13-SL100_BRANCH.orig/fs/nfsd/nfs3acl.c
+++ linux-2.6.13-SL100_BRANCH/fs/nfsd/nfs3acl.c
@@ -101,7 +101,7 @@ static int nfsd3_proc_setacl(struct svc_
 	int nfserr = 0;
 
 	fh = fh_copy(&resp->fh, &argp->fh);
-	nfserr = fh_verify(rqstp, &resp->fh, 0, MAY_NOP);
+	nfserr = fh_verify(rqstp, &resp->fh, 0, MAY_SATTR);
 
 	if (!nfserr) {
 		nfserr = nfserrno( nfsd_set_posix_acl(
